tcmur_cmd_handler.c

```diff
@@ -456,6 +456,134 @@ static int handle_read(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
 	return async_handle_cmd(dev, cmd, read_work_fn);
 }
 
+/* format unit */
+struct format_unit_state
+{
+	size_t length;
+	off_t offset;
+	void *write_buf;
+	struct tcmulib_cmd *origcmd;
+	uint32_t done_blocks;
+};
+
+static int format_unit_work_fn(struct tcmu_device *dev,
+			       struct tcmulib_cmd *writecmd)
+{
+	struct tcmulib_handler *handler = tcmu_get_dev_handler(dev);
+	struct tcmur_handler *rhandler = handler->hm_private;
+	struct tcmulib_cmd *origcmd = writecmd->cmdstate;
+	struct format_unit_state *state = origcmd->cmdstate;
+
+	return rhandler->write(dev, writecmd, writecmd->iovec,
+			       writecmd->iov_cnt, state->length, state->offset);
+}
+
+static void handle_format_unit_cbk(struct tcmu_device *dev,
+				   struct tcmulib_cmd *writecmd, int ret)
+{
+	struct tcmulib_cmd *origcmd = writecmd->cmdstate;
+	struct format_unit_state *state = origcmd->cmdstate;
+	uint8_t *sense = origcmd->sense_buf;
+
+	writecmd->iovec->iov_base = state->write_buf;
+	state->offset += state->length;
+	state->done_blocks += state->length / dev->block_size;
+	if (state->done_blocks < dev->num_lbas) {
+		if ((dev->num_lbas - state->done_blocks) * dev->block_size < state->length) {
+			state->length = (dev->num_lbas - state->done_blocks) * dev->block_size;
+			free_iovec(writecmd);
```

Comment:
Is the problem that tcmu_iovec_length would report a different length than state->length?

Make this a code comment. I do not think anyone is going to remember and when looking at the code it is not so clear right away.